### ISSUE TITLE
How to speed up SHAP computation

### ISSUE THREAD

**Issue by bingojojstu** [shap/shap#77]
Hi,

The package itself is really interesting and intuitive to use.
I notice however it takes quite long time to run on neural network with practical feature & sample size using KernelExplainer.
Question, is there any document to explain how to properly choose 
1. sample size fed into shap.KernelExplainer, and what is the guiding principal to choose these samples; 
2. number of samples fed into function explainer.shap_values, I would assume it has something to do with number of features(columns)

For example, I have over 1 million record with 400 raw features (continuous + unencoded categorical). 
Any suggestion would be appreciated.

![screen shot 2018-05-08 at 2 50 41 pm](https://user-images.githubusercontent.com/20424880/39785011-61aa402a-52cf-11e8-90d3-76ddddcddb8e.png)

Above screen shot is the example using 50 samples in KernelExplainer as typical feature values and 2000 case with 500 repeats in shap_values perturbation.

**Comment by slundberg** [comment]
Explaining models with large input spaces is challenging (i.e. your 400 features). Two suggestions I have are 1) either try using a smaller set of reference values (instead of 50). If for some reason there is a value that means "missing", such as 0, that will speed things up a lot (like bag of words).  2) Think if there are ways to group the model features together into units that could be treated together (like 1 hot groups, super pixels, etc.).

If that is not enough you could consider using a deep learning specific method such as DeepLIFT or integrated gradients. The output of DeepLIFT can be viewed as an approximation of SHAP values. Note both methods only handle a single reference value (rather than integrating over 50 like you are doing now).

Finally, not knowing the type of data it is hard to say, but you might also consider a GBM model like XGBoost since if your features are not correlated (pixels, sequences, etc.) GBMs are often state of the art, and Tree SHAP will explain instances very fast for GBMs.

**Comment by ybeybe** [comment]
@bingojojstu , I was wondering how to show the progress bar while running? is this something implemented in SHAP? Thanks.

**Comment by bingojojstu** [comment]
@ybeybe yes it is implemented in SHAP.

**Comment by slundberg** [comment]
FYI...the progress bar is only for KernelExplainer and SamplingExplainer. And for anyone else landing on this thread DeepExplainer and GradientExplainer are now good alternatives in SHAP if you are using a Neural network.

**Comment by athorneak13** [comment]
I'm trying to implement shap_values with XGBoost and it is still taking forever. I have 35 features and limited the number of samples to 500. I set the tree_limit to 10, but I don't really understand what that input means - doesn't XGBoost provide a single tree to use? Why would there be multiple trees used? Any advice on how to decrease the run time would be super helpful. 

Thanks

**Comment by slundberg** [comment]
Could you post the code you are using to explain the model? TreeExplainer
is usually very quick. XGBoost trains many trees (often thousands) so
`tree_limit` is useful if you want to explain the model with a different
early stopping criteria than was used while training the model.

**Comment by jarandaf** [comment]
I am somehow experiencing the same issue with a relatively large dataset (3M rows x 30 features), the `shap_values` method is taking ages. Is the library expected to be able to handle such sizes? The underlying model is LightGBM, and from what I read around here, gradient boosting models implementation should be fast.

**Comment by slundberg** [comment]
@jarandaf Fast is relative :) 3M is a lot of rows. If you post the runtime for 10k rows I can tell you if that sounds reasonable. Also if you have lots of trees in your model then the `approximate=True` option in TreeExplainer's shap_values method can give you fast approximate answers.

**Comment by astrixg** [comment]
Any update on @athorneak13's query regarding SHAP TreeExplainer performance? I'm facing the same issue right now.

Code snippet -
```
import shap
shap.initjs()
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X_train)
```
The data set is quite small (~150 data points), and the model training process was quick too. Is there a way to use parallel processing here? Its taking the same time on a 2 vCPUs 8 GB memory machine and an 8 vCPUs, 30 GB memory machine. I noticed that it's using only 1 core and around 2.5 GB memory in both cases.

**Comment by slundberg** [comment]
@astrixg Could you provide more details about the model? It seems like it should run quickly for only 150 samples.

**Comment by astrixg** [comment]
`model = XGBRegressor('colsample_bytree': 0.5, 'colsample_bylevel': 0.9, 'learning_rate': 0.5, 'min_child_weight': 1, 'subsample': 0.5, 'reg_lambda': 0.7, 'max_depth': 7, 'gamma': 0, 'n_estimators': 500)`

no. of features = 97

**Comment by astrixg** [comment]
@slundberg Previously the shap_values generation was taking around 8-10 mins, which has now dropped to around 4 mins. While I haven't made any changes to the data or the model, I did upgrade the SHAP version, but not sure if that is what caused the change.

**Comment by slundberg** [comment]
You can try just calling shap inside XGBoost directly with its `pred_contribs=True` option. But with the model you gave it should take less than a second to explain 150 samples, so even 4 min seems like something else is going on.

**Comment by zpconn** [comment]
I'm also finding `shap_values` to be prohibitively and unusably expensive on real-world datasets. It takes a day of runtime on 20 cores at 100% each to finish on a small dataset with only 3.9M rows (~150 features each). Considering many real-world datasets have 100M+ rows, this is actually far more expensive than training the models, and it could likely take literally months to run `shap_values` on all of our datasets.

To clarify, this is all for a tree-based GBT model.

**Comment by slundberg** [comment]
@zpconn how many trees are you using? That does seems a bit slower than I would expect. You might also try a batch of 100k and compare the per-sample runtime. I have not tested the performance on datasets with millions of samples and it could cause some memory-based slowdowns.

My other question is what do you plan to do with 100 million explanations? If you are just using summary statistics and plots then explaining just 10k will give plenty of detail. You can also try the approximate method if you have a lot of trees.

**Comment by dp-dpo** [comment]
Are there any parameters to control/force parallelization? "shap_values" seems to only load about 25% (=12 cores) of my CPU. 
I'm running a custom model with KernelExplainer (at about 1.5 it/s) and it basically takes forever (3 days), even though the predict takes only a second on its own. Is there any architectural way to make it work better with SHAP (apart from obviously making it simply predict faster)?

**Comment by bingojojstu** [comment]
reduce background sample size

**Comment by thomasmooon** [comment]
Hi @slundberg, great package, many thanks. 

## the case
I'm using a dataset with 12K Observations and ~ 160K features. The neural net has ~ 200K parameters. The walltime to perform   is described in the table below and I observed that it seems to grow exponentially with the number of samples (and not linear as described in the docstring of the `shap.Deepexplainer()` method). Using all 12K samples might take ~31days. Whilst monitoring the machine load I observed, that only 2-3 CPU cores (of 16) are used and GPUs weren't touched. Therefore I tried to run the job in a `tf.session()` with different number of cores, but that had no effect (see table below).

## the question
The only way I can currently imagine to get a robust estimate of the importancies is to perform `shap.Deepexplainer()` over multiple bootstraps for a good tradeoff between speed and accuracy. Do you have another / better idea? Thank you.

## walltime as a function of samplesize
size [%] | #samples | runtime [min] | runtime [days]
------|---------------------|----------------------|----------------------
1 | 122 | 1.6 (=95s)
5 | 610 | 66
100 | 12,218 | 44,262 (estimated)* | ~31d (estimated)*

> estimation:
> log(66/1.6)) / log(5) = 2.17
> 66 min * (100/5)^2.17 = ~31d

## walltime as a function of #CPUS (using 1% = 122 samples)

PARAMS['CPUS'] | CPUS used htop | runtime `shap.Deepexplainer()`
-- | -- | --
8 | 2 - 3 | 96 sec
1 | 2 - 3 | 95 sec

### related code (to try to force using *only* and *varying* no. of CPUs)
```
if PARAMS['CPUS']>0:
    import os
    os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"  
    os.environ["CUDA_VISIBLE_DEVICES"]=""
    
    import tensorflow as tf
    import keras
    config = tf.ConfigProto(device_count={"CPU": PARAMS['CPUS']})
    keras.backend.tensorflow_backend.set_session(tf.Session(config=config))
    sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
    print(sess)

```

**Comment by s-wxy** [comment]
> You can try just calling shap inside XGBoost directly with its `pred_contribs=True` option. But with the model you gave it should take less than a second to explain 150 samples, so even 4 min seems like something else is going on.

But, it looks like the `pred_contribs` only works for xgb.Booster. I trained an xgbClassifier model, on 1.1m+ data with 310 features. It takes 200min+ to get the shap values... and the `pred_contribs` not working for xgbClassifier ...

**Comment by christabella** [comment]
Are the following code snippets equivalent, and if so, couldn't we easily parallelize each iteration of the for loop in the first?
```
shap_values = []
for idx in range(num_rows):
    shap_val = explainer.shap_values(data.iloc[idx,:])
    shap_values.append(shap_val)
```
```
shap_values = explainer.shap_values(data.iloc[:num_rows,:])
```

**Comment by mirekphd** [comment]
> Explaining models with large input spaces is challenging

But such data shape is bound to be the most popular use case of SHAP - it is so good at feature selection:) SHAP is the only method I have found to be superior (yielding better metrics on new data for the same features number) to the previous state-of-the-art - the "split" variable importance from LightGBM, for selecting features for any boosted tree family model. I intend to run `GPUTreeExplainer` on 2k features and millions of rows routinely as soon as the GPU version starts working. 

I have seen some interesting results on real-life data for both regression and classification models (again, still using only samples for performance reasons), and the differences if favor of SHAP aren't small at all, in fact they are surprisingly large and tend to improve with sample size and increased realism, i.e. degree of model optimization. Xia Xiaomao et al. 2019 do not do it justice, by using unrealistic experimental setups, including toy datasets, inadequate in both dimensions, a suboptimal algorithm (XGB, RF instead of LGBM), and a poor evaluation method (single point at full feature set, where the diffs are smallest, instead of the whole curve across different feature set sizes). If you think @slundberg that the use of SHAP as a feature selector is novel enough to merit your academic interest, then I can email you those early results and some pointers on how to reproduce them on your datasets.

**Comment by rjp23** [comment]
I have a related issue, it's not a particularly large dataset (365 rows) that we're predicting but it is a large model trained on millions of data points:

```
XGBRegressor(base_score=0.5, booster='gbtree',
             colsample_bylevel=0.691731521840133,
             colsample_bynode=0.7164946223619781,
             colsample_bytree=0.8888804271481594, gamma=0.3264752658518928,
             gpu_id=-1, importance_type='gain', interaction_constraints='',
             learning_rate=0.055960860696856574, max_delta_step=0, max_depth=23,
             min_child_weight=1.6443597926092657, missing=nan,
             monotone_constraints='()', n_estimators=2150, n_jobs=1,
             num_parallel_tree=1, random_state=42, reg_alpha=0, reg_lambda=1,
             scale_pos_weight=1, subsample=0.6521107257366314,
             tree_method='exact', validate_parameters=1, verbosity=2)
```

The model loads in seconds:

```
model_file = 'saved_model.xgb'
model = XGBRegressor()
model.load_model(model_file)
```

but TreeExplainer takes a LONG time (hours, if successful at all)
`explainer = shap.TreeExplainer(model)
`

A smaller version of the model (trained on less data) does work in a reasonable amount of time so I think it's just due to the model being very large. Is there any way to optimise TreeExplainer for it?

**Comment by VEZcoding** [comment]
Hey! I have a model of XGBoost Classifier and a matrix of 130 samples and around 25000 features. It takes seconds for a TreeExplainer finishes. Am I doing something wrong?

**Comment by csetzkorn** [comment]
> Hey! I have a model of XGBoost Classifier and a matrix of 130 samples and around 25000 features. It takes seconds for a TreeExplainer finishes. Am I doing something wrong?

I think having 25000 features but only 130 samples could be deemed wrong in the first place? Unless I miss something  ...

**Comment by avibrazil** [comment]
`shap.TreeExplainer.shap_values()` takes forever to execute while using only 1 CPU, leaving my other 19 CPUs idle.
The code in this method needs to be parallelized.

**Comment by imiordanov** [comment]
I second this. I'm running SHAP on a rather large dataset, and it loos like it's using only 1 of the 12 CPU cores that I have available. With the current computation speed, it will take around one full month of continuous computations just to obtain the SHAP values.

**Comment by pmoriano** [comment]
I see this issue is still open. Many thanks to @slundberg for his amazing contribution with `SHAP`. 

I am trying to get `shap_values` from a `model=RandomForestClassifier(n_estimators=1000)` (all the other parameters are the default). The model was trained on a dataset of dimension `(47602, 23)`. I am using this model to compute `shap_values` on a test dataset of dimension `(20402, 23)`, but it takes forever (more than five hours by now). I am using `shap` version `0.41.0`.

I wonder if as of today, my situation looks reasonable and if there are suggestions to speedup the computation. I saw that the`GPUTree` is there, but wanted to double check if this is the most suggested route to proceed and if there are sample code available about how to use it. BTW, the code I am using is below:

```
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X_test) 
```

Thanks for the help.

**Comment by avibrazil** [comment]
A theoretical solution would be to split your dataset by the number of CPUs you have and run shap_values() on each CPU to compute shap for each data partition. Then concatenate all the results together.

I was ready to implement this in my app but then I've found the [fasttreeshap](https://github.com/linkedin/FastTreeSHAP) module which incorporates a few optimizations, including probably this one too.

**Comment by ricardobarroslourenco** [comment]
@avibrazil have you tested any of the in implementations of FastSHAP (https://openreview.net/forum?id=Zq2G_VTV53T ) with PyTorch or TF? They claim speedups of 200-2000x, but I wonder if it is comparable in terms of experimental reproducibility.

**Comment by arilwan** [comment]
I am still having this reported issue. In my case, the model was built on train set size: `(67092, 80)` and the test I am explaining is shaped: `(40682, 80)`.

This way:
```python
explainer = shap.TreeExplainer(model)
shap_values =  explainer.shap_values(X_test)
```

As at now, this takes 5 days and still running, on a 16 core CPU.

Apparently, this is NOT taking advantage of parallelism. Any workaround for speed up?

**Comment by talismanbrandi** [comment]
> @zpconn how many trees are you using? That does seems a bit slower than I would expect. You might also try a batch of 100k and compare the per-sample runtime. I have not tested the performance on datasets with millions of samples and it could cause some memory-based slowdowns.
> 
> My other question is what do you plan to do with 100 million explanations? If you are just using summary statistics and plots then explaining just 10k will give plenty of detail. You can also try the approximate method if you have a lot of trees.

There are genuine use cases for computing Shapley values for O(10M) samples. We are doing so to build interaction networks of proteins and RNAs. Instead of protein binding data, we are using local Shapley values. There is a way to do it with pySpark: https://www.databricks.com/blog/2022/02/02/scaling-shap-calculations-with-pyspark-and-pandas-udf.html in case anyone needs it.

**Comment by panoanx** [comment]
I'm also facing similar issues. Single threading is the main bottleneck when dealing with a large bunch of data samples. 
So it's natural to accelerate the calculation using multiprocessing for parallelization. Here is my implementation:

```python
import multiprocessing as mp


def infer_shap_values(X_tensor, batch_size): # X_tensor: shape(num_samples, num_features)
    explainer = shap.GradientExplainer(mlp_model, X_tensor, batch_size=batch_size)
    shap_values = []

    ctx = mp.get_context("spawn")
    with ctx.Pool(processes=N_PROC) as pool:
        args = [
            (X_tensor[start_idx:end_idx], batch_size)
            for start_idx, end_idx in zip(
                range(0, len(X_tensor), batch_size),
                range(batch_size, len(X_tensor) + batch_size, batch_size),
            )
        ]
        shap_values_list = pool.starmap(explainer.shap_values, args)
        shap_values = np.concatenate([np.array(sv) for sv in shap_values_list])

    return shap_values


shap_values = infer_shap_values(Some_X, batch_size=50)
```

I tested with a simple PyTorch MLP. For 1000 samples it takes about ~50 seconds on 1 cpu core, and ~10 seconds on 8 cpu cores, i.e. almost 5x faster.
